### PR TITLE
NMS-14049: Use poller definition for device config backup

### PR DESCRIPTION
**Pull request by fooker**
### External References

* JIRA (Issue Tracker): http://issues.opennms.org/browse/NMS-14049

### CODE REVIEW

File: features/device-config/api/src/main/java/org/opennms/features/deviceconfig/service/DeviceConfigService.java [hunk 1]
```diff
@@ -38,20 +39,28 @@
...
      * @param timeout    timeout in milliseconds for retrieving device config
      * @throws IOException
      * @return
      */
-    CompletableFuture<byte[]> getDeviceConfig(String ipAddress, String location, String configType, int timeout) throws IOException;
+    CompletableFuture<byte[]> getDeviceConfig(String ipAddress, String location, String service, int timeout) throws IOException;
+
+    List<RetrivalConfiguration> getRetrivalConfigurations(String ipAddress, String location);
+
+    interface RetrivalConfiguration {
```
Comment (synqotik) [inline features/device-config/api/src/main/java/org/opennms/features/deviceconfig/service/DeviceConfigService.java:61]: Here and elsewhere: `Retrival` -> `Retrieval`

File: opennms-config/src/main/java/org/opennms/netmgt/config/PollerConfig.java [hunk 2]
```diff
@@ -295,7 +296,23 @@
     public List<Package> getPackages();
 
     List<Monitor> getConfiguredMonitors();
-    
+
+    /**
+     * Find the {@link Package} containing the service selevted for the given IP.
+     * @param ipAddr the address to select the package for
```
Comment (synqotik) [inline opennms-config/src/main/java/org/opennms/netmgt/config/PollerConfig.java:302]: selevted -> selected

File: features/device-config/service/src/main/java/org/opennms/features/deviceconfig/service/impl/DeviceConfigServiceImpl.java [hunk 3]
```diff
@@ -112,20 +114,69 @@ public void triggerConfigBackup(String ipAddress, String location, String config
...
+
+        return iface
+                // Get all device config services defined for this interface
+                .getMonitoredServices().stream()
+
+                // Resolve the service name into service config
+                .flatMap(svc -> this.pollerConfig.findService(InetAddressUtils.str(svc.getIpAddress()), svc.getServiceName()).stream())
+
+                // Filter for the device config monitor
+                .filter(match -> this.pollerConfig.getServiceMonitor(match.service.getName()).getClass() == DeviceConfigMonitor.class)
```
Comment (cgorantla) [inline features/device-config/service/src/main/java/org/opennms/features/deviceconfig/service/impl/DeviceConfigServiceImpl.java:129]: This brings direct dependency on DeviceConfigMonitor for the service.  
One messy way is to get Monitors by className from `PollerRequestBuilderImpl` just like you did with `getInterpolatedAttributes()` and match the monitor.

File: features/device-config/api/src/main/java/org/opennms/features/deviceconfig/service/DeviceConfigService.java [hunk 4]
```diff
@@ -38,20 +39,28 @@
...
      * @param ipAddress  specific IpAddress for which we need to fetch device config.
      * @param location   specific minion location at which we need to fetch device config.
-     * @param configType configType whether it is Default or Running.
+     * @param service    name of the bound service.
      * @param timeout    timeout in milliseconds for retrieving device config
      * @throws IOException
      * @return
      */
-    CompletableFuture<byte[]> getDeviceConfig(String ipAddress, String location, String configType, int timeout) throws IOException;
+    CompletableFuture<byte[]> getDeviceConfig(String ipAddress, String location, String service, int timeout) throws IOException;
```
Comment (cgorantla) [inline features/device-config/api/src/main/java/org/opennms/features/deviceconfig/service/DeviceConfigService.java:57]: Need to persist the service name in DB.

File: features/device-config/service/src/main/java/org/opennms/features/deviceconfig/service/impl/DeviceConfigServiceImpl.java (hunk 3 above)
Comment (fooker) [reply features/device-config/service/src/main/java/org/opennms/features/deviceconfig/service/impl/DeviceConfigServiceImpl.java:129]: Any reason why we should avoid this dependency? Having it makes it save for refactoring.

I can't see how the `PollerRequestBuilder` could help here. This is about resolving the service name to the assigned monitoring class, which is only available from poller config.

File: features/device-config/api/src/main/java/org/opennms/features/deviceconfig/service/DeviceConfigService.java (hunk 1 above)
Comment (fooker) [reply features/device-config/api/src/main/java/org/opennms/features/deviceconfig/service/DeviceConfigService.java]: fixed

File: opennms-config/src/main/java/org/opennms/netmgt/config/PollerConfig.java (hunk 2 above)
Comment (fooker) [reply opennms-config/src/main/java/org/opennms/netmgt/config/PollerConfig.java:302]: changed

File: features/device-config/api/src/main/java/org/opennms/features/deviceconfig/service/DeviceConfigService.java (hunk 4 above)
Comment (fooker) [reply features/device-config/api/src/main/java/org/opennms/features/deviceconfig/service/DeviceConfigService.java:57]: Done

**Comment by fooker** [comment]
@synqotik this also includes a change to the rest API. Is there any need for sync up with the frontend?

**Comment by fooker** [comment]
There is a `smoke` branch, too: https://app.circleci.com/pipelines/github/OpenNMS/opennms?branch=jira%2FNMS-14049-smoke&filter=all

File: features/device-config/rest/pom.xml [hunk 5]
```diff
@@ -52,6 +52,11 @@
             <artifactId>commons-compress</artifactId>
             <version>${commonsCompressVersion}</version>
         </dependency>
+        <dependency>
```
[the remark this answers is not in the document]
Comment (fooker) [reply features/device-config/rest/pom.xml:55]: Changed. But impl is still required for tests.

**Review by christianpape — APPROVED** [review]
LGTM.

**Review by cgorantla — COMMENTED** [review]
LGTM!
Will give it a whirl tomorrow.

**Pull request merged**